include/envoy/network/io_handle.h

```diff
@@ -65,6 +66,15 @@ class IoHandle {
   virtual Api::IoCallUint64Result readv(uint64_t max_length, Buffer::RawSlice* slices,
                                         uint64_t num_slice) PURE;
 
+  /**
+   * Read from a io handle directly into buffer.
+   * @param buffer supplies the buffer to read into.
+   * @param max_length supplies the maximum length to read.
+   * @return a IoCallUint64Result with err_ = nullptr and rc_ = the number of bytes
+   * read if successful, or err_ = some IoError for failure. If call failed, rc_ shouldn't be used.
+   */
+  virtual Api::IoCallUint64Result read(Buffer::Instance& buffer, uint64_t max_length) PURE;
```

Comment:
Moving read to IoHandle makes sense to me since some IoHandles may be able to move data to the buffer via mechanisms more efficient than reserve + readv + commit.  The increase in the number of calls to Buffer::reserve outside the buffer class is unfortunate, but I think the benefits of this API change out-weight the drawbacks.